tests/strings/test_z_algorithm.py

```diff
@@ -25,14 +22,26 @@ def generate_words(char_set, length):
 
 def test_z_algorithm__corner_cases():
```

Comment:
I really think the one test we should have is 
```
n = 1 << 10
for i in range(n):
    S = bin(i + n)[2:]
    assert ...
```
Maybe even n = 1 << 15. This test will catch a ton of possible bugs, we should probably have a test like this for every string algorithm. Maybe we could also add a test for all possible strings of length 7 of 3 characters.

Reply:
I don't understand. We already have a test called `test_z_algorithm__corner_cases`, which generates all possible strings of length 1-10 with the first 1-3 characters. Why do we need this test you are suggesting?

Reply:
Ah yes, sorry. Missed that.

Btw your for loop over c is currently unnecessary, you could just do 
```
def test_z_algorithm__corner_cases():
    for n in range(1, 11):
        for word in product('abcd', repeat=n):
            assert z_function(word) == [_get_lcp(word, i) for i in range(n)]
```
The product call with 'abcd' already contains strings like 'aaaa', so no need to have the c for loop.

Reply:
How about now?

Yes, it is not necessary to join the tuple. However I'd like to have it explicit, since usually the function will be called with a string. However due to the way it's implemented it will also work for lists and tuples.

Reply:
Having `word = ''.join(c for c in word)` is unnecessary. Just do `word = ''.join(word)`. 

Still don't really get  why you want to join it in the first place. In what way does a tuple of characters differ from a string? The only difference I could think of is if you want to print it. I believe all string algorithms in Pyrival should run for input of strings or lists of integers or tuple of characters etc.